In [1]:
import pandas as pd
import numpy as np

In [2]:
column_names = ['id_number', 'psa_level', 'cancer_volume', 'weight', 'age', 
                'benign_prostatic_hyperplasia', 'seminal_vesicle_invasion', 
                'capsular_penetration', 'gleason_score']

In [8]:
data = pd.read_csv('../resources/assignment3_1_cpy.txt', sep = ',', header = None, names = column_names)

In [10]:
data.describe()

id_number   psa_level  cancer_volume      weight        age  \
count  97.000000   97.000000      97.000000   97.000000  97.000000   
mean   49.000000   23.730134       6.998682   45.491361  63.865979   
std    28.145456   40.782925       7.880869   45.705053   7.445117   
min     1.000000    0.651000       0.259200   10.697000  41.000000   
25%    25.000000    5.641000       1.665300   29.371000  60.000000   
50%    49.000000   13.330000       4.263100   37.338000  65.000000   
75%    73.000000   21.328000       8.414900   48.424000  68.000000   
max    97.000000  265.072000      45.604200  450.339000  79.000000   

       benign_prostatic_hyperplasia  seminal_vesicle_invasion  \
count                     97.000000                 97.000000   
mean                       2.534725                  0.216495   
std                        3.031176                  0.413995   
min                        0.000000                  0.000000   
25%                        0.000000                  0.000000   
50%                        1.349900                  0.000000   
75%                        4.758800                  0.000000   
max                       10.277900                  1.000000   

       capsular_penetration  gleason_score  
count             97.000000      97.000000  
mean               2.245367       6.876289  
std                3.783329       0.739619  
min                0.000000       6.000000  
25%                0.000000       6.000000  
50%                0.449300       7.000000  
75%                3.254400       7.000000  
max               18.174100       8.000000

In [12]:
predictors = ['cancer_volume', 'weight', 'age', 
                'benign_prostatic_hyperplasia', 'seminal_vesicle_invasion', 
                'capsular_penetration', 'gleason_score']

In [13]:
response = ['psa_level']

### data preprocessing

In [21]:
data.iloc[0:5]

id_number  psa_level  cancer_volume  weight  age  \
0          1      0.651         0.5599  15.959   50   
1          2      0.852         0.3716  27.660   58   
2          3      0.852         0.6005  14.732   74   
3          4      0.852         0.3012  26.576   58   
4          5      1.448         2.1170  30.877   62   

   benign_prostatic_hyperplasia  seminal_vesicle_invasion  \
0                           0.0                         0   
1                           0.0                         0   
2                           0.0                         0   
3                           0.0                         0   
4                           0.0                         0   

   capsular_penetration  gleason_score  
0                   0.0              6  
1                   0.0              7  
2                   0.0              7  
3                   0.0              6  
4                   0.0              6

In [14]:
X = data[predictors]

In [19]:
X = X.values

In [20]:
X

array([[  5.59900000e-01,   1.59590000e+01,   5.00000000e+01,
          0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
          6.00000000e+00],
       [  3.71600000e-01,   2.76600000e+01,   5.80000000e+01,
          0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
          7.00000000e+00],
       [  6.00500000e-01,   1.47320000e+01,   7.40000000e+01,
          0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
          7.00000000e+00],
       [  3.01200000e-01,   2.65760000e+01,   5.80000000e+01,
          0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
          6.00000000e+00],
       [  2.11700000e+00,   3.08770000e+01,   6.20000000e+01,
          0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
          6.00000000e+00],
       [  3.49900000e-01,   2.52800000e+01,   5.00000000e+01,
          0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
          6.00000000e+00],
       [  2.09590000e+00,   3.21370000e+01,   6.40000000e+01,
          1.85890000e+00,   0.00

In [22]:
y = data[response]

In [24]:
y = y.values

In [38]:
x_tst = np.array([4.2633, 22.783, 68, 1.3500, 0, 0, 6])

In [39]:
x_tst

array([  4.2633,  22.783 ,  68.    ,   1.35  ,   0.    ,   0.    ,   6.    ])

### Build Models

#### Try Linear Regression

In [27]:
import statsmodels.api as sm
from statsmodels.sandbox.regression.predstd import wls_prediction_std

/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [32]:
result = sm.OLS(y,X).fit()

In [33]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.596
Model:                            OLS   Adj. R-squared:                  0.564
Method:                 Least Squares   F-statistic:                     18.96
Date:                Tue, 10 Apr 2018   Prob (F-statistic):           2.50e-15
Time:                        00:09:56   Log-Likelihood:                -467.16
No. Observations:                  97   AIC:                             948.3
Df Residuals:                      90   BIC:                             966.3
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1             2.0549      0.588      3.496      0.001       0.887       3.222
x2             0.0091      0.073      0.124      0.902      -0.137       0.155
x3            -0.6362      0.394     -1.613      0.110      -1.420       0.147
x4             1.4053      1.162      1.209      0.230      -0.903       3.714
x5            20.2485     10.707      1.891      0.062      -1.023      41.520
x6             1.1540      1.319      0.875      0.384      -1.467       3.775
x7             5.6900      3.687      1.543      0.126      -1.634      13.014
==============================================================================
Omnibus:                       95.209   Durbin-Watson:                   0.822
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1166.316
Skew:                           3.126   Prob(JB):                    5.47e-254
Kurtosis:                      18.795   Cond. No.                         289.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [40]:
predictions = result.get_prediction(x_tst)

In [41]:
predictions.summary_frame(alpha=0.05)

mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  1.739799  6.264748     -10.706216      14.185815    -61.128654   

   obs_ci_upper  
0     64.608252

#### Result is bad (Only 0.564 adjusted R-Square)

#### Try standardization for inputs

In [36]:
from sklearn import preprocessing

In [80]:
scaler = preprocessing.StandardScaler().fit(X)
scaler_y = preprocessing.StandardScaler().fit(y)

In [57]:
scaler.mean_

array([  6.99868247,  45.49136082,  63.86597938,   2.53472474,
         0.21649485,   2.24536701,   6.87628866])

In [59]:
x_scaled = scaler.transform(X)

In [62]:
x_scaled.mean(axis=0)

array([ -1.37347178e-17,   2.06020767e-16,   4.13186095e-16,
        -4.57823928e-17,  -3.66259142e-17,  -4.57823928e-17,
        -6.36375259e-16])

In [82]:
y_scaled = scaler_y.transform(y)

In [98]:
result_after_rescale = sm.OLS(y,X[:,[1,2,3,6]]).fit()

In [97]:
X[:,[1,2,3,6]]

array([[  1.59590000e+01,   5.00000000e+01,   0.00000000e+00,
          6.00000000e+00],
       [  2.76600000e+01,   5.80000000e+01,   0.00000000e+00,
          7.00000000e+00],
       [  1.47320000e+01,   7.40000000e+01,   0.00000000e+00,
          7.00000000e+00],
       [  2.65760000e+01,   5.80000000e+01,   0.00000000e+00,
          6.00000000e+00],
       [  3.08770000e+01,   6.20000000e+01,   0.00000000e+00,
          6.00000000e+00],
       [  2.52800000e+01,   5.00000000e+01,   0.00000000e+00,
          6.00000000e+00],
       [  3.21370000e+01,   6.40000000e+01,   1.85890000e+00,
          6.00000000e+00],
       [  3.44670000e+01,   5.80000000e+01,   4.66460000e+00,
          6.00000000e+00],
       [  3.44670000e+01,   4.70000000e+01,   0.00000000e+00,
          7.00000000e+00],
       [  2.55340000e+01,   6.30000000e+01,   0.00000000e+00,
          6.00000000e+00],
       [  3.65980000e+01,   6.50000000e+01,   0.00000000e+00,
          6.00000000e+00],
       [  3.65980000e

In [99]:
result_after_rescale.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.354
Model:                            OLS   Adj. R-squared:                  0.326
Method:                 Least Squares   F-statistic:                     12.72
Date:                Tue, 10 Apr 2018   Prob (F-statistic):           2.70e-08
Time:                        00:45:40   Log-Likelihood:                -489.95
No. Observations:                  97   AIC:                             987.9
Df Residuals:                      93   BIC:                             998.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1             0.0370      0.091      0.406      0.686      -0.144       0.218
x2            -1.3729      0.473     -2.900      0.005      -2.313      -0.433
x3             0.4905      1.434      0.342      0.733      -2.358       3.339
x4            15.9017      4.225      3.764      0.000       7.512      24.291
==============================================================================
Omnibus:                      117.862   Durbin-Watson:                   0.297
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2000.898
Skew:                           4.181   Prob(JB):                         0.00
Kurtosis:                      23.619   Cond. No.                         91.9
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [65]:
result_after_rescale.rsquared()

TypeError: 'numpy.float64' object is not callable

#### Try SVM

In [72]:
from sklearn.svm import SVR
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_validate

In [73]:
svm_clf = SVR(kernel = 'linear')

In [74]:
scoring = ['r2']

In [86]:
scores = cross_validate(svm_clf, x_scaled, y_scaled, cv=10, scoring = scoring, return_train_score=False)

/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/vali

In [87]:
scores

{'fit_time': array([ 0.00676203,  0.0027771 ,  0.00151706,  0.00147915,  0.00136971,
         0.00295186,  0.00340295,  0.00555182,  0.00277901,  0.00365996]),
 'score_time': array([ 0.00155997,  0.00033784,  0.00045013,  0.00042605,  0.00030518,
         0.00026011,  0.00155902,  0.00031519,  0.00132298,  0.00028706]),
 'test_r2': array([ -32.77266984, -212.98937311, -172.15961968, -794.22410161,
         -41.85982573, -242.32459271, -204.72320276,  -15.53508027,
         -49.96108831,   -1.19963875])}

#### Try random forest

In [101]:
from sklearn.ensemble import RandomForestRegressor

In [111]:
rf_clf = RandomForestRegressor(n_estimators = 100)

In [112]:
scores = cross_validate(svm_clf, x_scaled, y, cv=10, scoring = scoring, return_train_score=False)

/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/validation.py:547: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/xinsongdu/.pyenv/versions/miniconda3-latest/lib/python3.6/site-packages/sklearn/utils/vali

In [113]:
scores['test_r2']

array([ -34.73223974, -118.32181987, -164.4001281 , -422.65078554,
        -47.78181048, -126.59031698, -131.0603409 ,  -10.27852347,
        -46.93206541,   -1.2711713 ])